internal/nodepingcontroller/node_ping_controller.go

```diff
@@ -0,0 +1,136 @@
+package nodepingcontroller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/virtual-kubelet/virtual-kubelet/lock"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	"github.com/virtual-kubelet/virtual-kubelet/trace"
+	vktypes "github.com/virtual-kubelet/virtual-kubelet/types"
+	"golang.org/x/sync/singleflight"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type NodePingController interface {
+	Run(ctx context.Context)
+	GetResult(ctx context.Context) (*PingResult, error)
+}
+
+type nodePingController struct {
+	nodeProvider vktypes.NodeProvider
+	pingInterval time.Duration
+	pingTimeout  *time.Duration
+	cond         lock.Cond
+
+	// "Results"
+	result *PingResult
+}
+
+type PingResult struct {
+	PingTime time.Time
+	Error    error
+}
+
+func NewNodePingController(node vktypes.NodeProvider, pingInterval time.Duration, timeout *time.Duration) NodePingController {
+	if pingInterval == 0 {
+		panic("Node ping interval is 0")
+	}
+
+	if timeout != nil && *timeout == 0 {
+		panic("Node ping timeout is 0")
+	}
+
+	return &nodePingController{
+		nodeProvider: node,
+		pingInterval: pingInterval,
+		pingTimeout:  timeout,
+		cond:         lock.NewCond(),
+	}
+}
+
+func (npc *nodePingController) Run(ctx context.Context) {
+	const key = "key"
+	sf := &singleflight.Group{}
+
+	// 1. If the node is "stuck" and not responding to pings, we want to set the status
+	//    to that the node provider has timed out responding to pings
+	// 2. We want it so that the context is cancelled, and whatever the node might have
+	//    been stuck on uses context so it might be unstuck
+	// 3. We want to retry pinging the node, but we do not ever want more than one
+	//    ping in flight at a time.
```

Comment:
This comment is concise and clear but the code to do it (cond, ticker, waiter) requires a bit of mental effort to follow, for a dumb programmer such as myself.

Reply:
Spent some more time w/ said code while reviewing #913. It makes more sense now. Hopefully, having that reviewed and, eventually, merged will reduce the amount of code in this PR.